feat: 【APM / Trace检索】候选项接口 --story=123958613

yamltdg opened the pull request.

yamltdg committed:
feat: 【APM / Trace检索】候选项接口&topk接口零值处理 #1010158081123958613

feat: 【APM / Trace检索】topk查询值排序兜底 #1010158081123958613

yamltdg commented:
测试
1. 指标表
1.1 trace 
request
```json
{
  "app_name": "bkop",
  "start_time": 1747152728,
  "end_time": 1747156328,
  "bk_biz_id": 2,
  "fields":["span_name"],
  "datasource_type":"metric",
  "limit":100,
  "query_string":""
}
```
response
```
{
    "result": true,
    "code": 200,
    "message": "OK",
    "data": {
        "span_name": [
            "/query/ts",
            "/query/ts/info/tag_values",
            "/query/ts/promql",
            "Elasticsearch request",
            "GET",
            "HTTP POST",
            "SELECT",
            "SET",
            "UPDATE",
            "monitor_web.grafana.resources.unify_query.GraphUnifyQueryResource",
            "monitor_web.grafana.resources.unify_query.UnifyQueryRawResource",
            "monitor_web.incident.resources.IncidentTopologyResource"
        ]
    }
}
```
1.2 span
request
```json
{
  "app_name": "bkop",
  "start_time": 1747152728,
  "end_time": 1747156328,
  "bk_biz_id": 2,
  "fields":["span_name"],
  "mode":"span",
  "datasource_type":"metric",
  "limit":10,
  "query_string":""
}
```
response
```json
{
    "result": true,
    "code": 200,
    "message": "OK",
    "data": {
        "span_name": [
            "/query/ts",
            "/query/ts/info/tag_values",
            "/query/ts/promql",
            "Elasticsearch request",
            "GET",
            "HTTP POST",
            "SELECT",
            "SET",
            "UPDATE",
            "monitor_web.grafana.resources.unify_query.GraphUnifyQueryResource",
            "monitor_web.grafana.resources.unify_query.UnifyQueryRawResource"
        ]
    }
}
```
2.trace表
2.1.trace
request
```json
{
  "app_name": "bkop",
  "start_time": 1747152728,
  "end_time": 1747156328,
  "mode":"trace",
  "bk_biz_id": 2,
  "fields":["root_service"],
  "datasource_type":"trace",
  "limit":10,
  "query_string":""
}
```
response
```json
{
    "result": true,
    "code": 200,
    "message": "OK",
    "data": {
        "root_service": [
            "",
            "cmdb_apiserver",
            "cmdb_coreservice",
            "cmdb_auth",
            "cmdb_topo",
            "cmdb_cacheservice",
            "cmdb_host",
            "unify-query",
            "cmdb_sidecar",
            "cmdb_proc"
        ]
    }
}
```
2.2 .span
request
```json
{
  "app_name": "bkop",
  "start_time": 1747152728,
  "end_time": 1747156328,
  "bk_biz_id": 2,
  "fields":["span_id"],
  "mode":"span",
  "datasource_type":"trace",
  "limit":10,
  "query_string":""
}
```
response
```json
{
    "result": true,
    "code": 200,
    "message": "OK",
    "data": {
        "span_id": [
            "00000b86ef7be978",
            "0000123b0138f271",
            "000016ef7aa7aa31",
            "00003a9a6c155314",
            "0000453bab1a8688",
            "00004bbd004e8862",
            "00004d6877d70d77",
            "000050bc581ce93e",
            "00005a9ee39eaf3a",
            "00005b69ad99c0d6"
        ]
    }
}
```

yamltdg commented on bkmonitor/apm/resources.py:
```diff
@@ -1936,11 +1937,12 @@ def perform_request(self, validated_data):
-                        "value": field_topk.get("field_value", ""),
-                        "count": int(field_topk.get("count", 0)),
-                        "proportions": round(100 * (field_topk.get("count", 0) / total), 2) if total > 0 else 0,
+                        "value": field_topk["field_value"],
```
unify-query会补0值，因此一定有数据，可以不用get

yamltdg commented on bkmonitor/apm/core/handlers/query/base.py:
```diff
@@ -385,7 +385,12 @@ def _query_field_topk(self, start_time, end_time, field, limit, filters: list[ty
+        except Exception as e:
+            logger.warning("failed to query field %s topk, error: %s", field, e)
+            raise ValueError(_(f"字段 {field} topk值查询出错"))
+        return sorted(field_topk_values, key=lambda topk_item: topk_item["_result_"], reverse=True)
```
查询数据排序兜底

ZhuoZhuoCrayon commented on bkmonitor/apm/core/handlers/query/base.py:
```diff
@@ -427,6 +432,37 @@ def query_graph_config(self, start_time, end_time, field, filters: list[types.Fi
+                for field in fields
+            ]
+        )
+        return {field: values for field, values in option_values.items()}
```
似乎等价？`return option_values`。

ZhuoZhuoCrayon commented on bkmonitor/apm/resources.py:
```diff
@@ -2203,3 +2205,39 @@ def collect_interval_buckets(cls, base_query_params, proxy, bucket, interval: tu
+        )
+
+
+class NewQueryTraceOptionValuesResource(Resource):
```
从代码逻辑上看，之前 `QuerySpanOptionValuesResource` 和 `QueryTraceOptionValuesResource` 能满足扩展 `filters` & `query_string` 的需求，不需要额外新开接口？

ZhuoZhuoCrayon commented on bkmonitor/apm/core/handlers/query/base.py:
```diff
@@ -427,6 +432,37 @@ def query_graph_config(self, start_time, end_time, field, filters: list[types.Fi
         )
         return self.time_range_queryset(start_time, end_time).add_query(q).instant().time_agg(False).config
 
+    def query_fields_option_values(
```
同理，逻辑合并到 `_query_option_values`。

ZhuoZhuoCrayon commented on bkmonitor/apm/core/handlers/query/base.py:
```diff
@@ -427,6 +432,37 @@ def query_graph_config(self, start_time, end_time, field, filters: list[types.Fi
+        fields: list[str],
+        limit: int,
+        filters: list[types.Filter],
+        query_string: str,
```
这里是不是缺少一个逻辑，如果 query_string 不是 `"*"` or `""`，或者 `filters` 存在的情况下，直接走 `LOG`？

yamltdg replied:
这里saas侧做了处理就可以知道datasource_type，后台通过datasource_type就可以直接确认走什么了，好像不需要再判断 query_string 或者filters了

ZhuoZhuoCrayon commented on bkmonitor/apm/core/handlers/query/base.py:
```diff
@@ -427,6 +432,37 @@ def query_graph_config(self, start_time, end_time, field, filters: list[types.Fi
+        query_string: str,
+    ) -> dict[str, list[str]]:
+        q: QueryConfigBuilder = (
+            self._get_q(datasource_type).filter(self._build_filters(filters)).query_string(query_string)
```
加上 `order_by("-_value")`，确保有序。

yamltdg committed:
feat: 【APM / Trace检索】字段候选项复用旧版option接口#1010158081123958613

fix: 【APM / Trace检索】删除无用的查询函数#1010158081123958613

fix: 【APM / Trace检索】代码优化#1010158081123958613

fix: 【APM / Trace检索】指标表查询条件问题修复#1010158081123958613

ZhuoZhuoCrayon commented on bkmonitor/apm/resources.py:
```diff
@@ -1030,12 +1033,31 @@ class QueryTraceOptionValues(Resource):
+                    "value": validated_data["app_name"],
+                    "operator": FilterOperator.EQUAL,
+                }
+            )
```
span_query / trace_query 保留各自的  `query_option_values` 方法，可以收敛掉这个分支，用 `self.build_app_filter()` 进行过滤。

yamltdg committed:
fix: 【APM / Trace检索】收敛Trace指标查询条件分支#1010158081123958613

ZhuoZhuoCrayon reviewed (approved):
LGTM
